tokens/usecase/tokens_usecase.go

```diff
@@ -349,40 +387,29 @@ func (t *tokensUseCase) getPricesForBaseDenom(ctx context.Context, baseDenom str
 }
 
 func (t *tokensUseCase) getChainScalingFactorMut(precision int) (osmomath.Dec, bool) {
-	result, ok := t.precisionScalingFactorMap[precision]
-	return result, ok
-}
-
-// GetTokensFromChainRegistry fetches the tokens from the chain registry.
-// It returns a map of tokens by chain denom.
-func GetTokensFromChainRegistry(chainRegistryAssetsFileURL string) (map[string]domain.Token, error) {
-	// Fetch the JSON data from the URL
-	response, err := http.Get(chainRegistryAssetsFileURL)
-	if err != nil {
-		return nil, err
+	result, ok := t.precisionScalingFactorMap.Load(precision)
+	if !ok {
+		return osmomath.Dec{}, false
 	}
-	defer response.Body.Close()
 
-	// Decode the JSON data
-	var assetList AssetList
-	err = json.NewDecoder(response.Body).Decode(&assetList)
-	if err != nil {
-		return nil, err
+	v, ok := result.(osmomath.Dec)
+	if !ok {
+		return osmomath.Dec{}, false
 	}
 
-	tokensByChainDenom := make(map[string]domain.Token)
+	return v, true
+}
 
-	// Iterate through each asset and its denom units to print exponents
-	for _, asset := range assetList.Assets {
-		token := domain.Token{}
-		token.Precision = asset.Decimals
-		token.HumanDenom = asset.Symbol
-		token.IsUnlisted = asset.Preview
-		token.CoingeckoID = asset.CoingeckoID
-		tokensByChainDenom[asset.CoinMinimalDenom] = token
+// UpdateAssetsAtHeightIntervalAsync updates assets at configured height interval.
+// Internally, it calls LoadTokensFromChainRegistry as a goroutine.
+func (t *tokensUseCase) UpdateAssetsAtHeightIntervalAsync(height uint64) {
+	if height%uint64(t.updateAssetsHeightInterval) == 0 {
+		go func() {
+			if err := t.tokenLoader.FetchAndUpdateTokens(t.LoadTokens); err != nil {
+				t.logger.Error("error loading tokens from chain registry", zap.Error(err))
+			}
+		}()
 	}
-
-	return tokensByChainDenom, nil
 }
```

Comment:
Consider simplifying to the following:
```go
// UpdateAssetsAtHeightIntervalAsync updates assets at configured height interval.
// Internally, it calls LoadTokensFromChainRegistry as a goroutine.
func (t *tokensUseCase) UpdateAssetsAtHeightIntervalAsync(height uint64) {
	if height%uint64(t.updateAssetsHeightInterval) == 0 {
		go func() {
			if err := t.updateAssetsAtHeightIntervalSync(height); err != nil {
				// TODO: add prometheus error counter so that we can create an alert for observability
				t.logger.Error("error loading tokens from chain registry", zap.Error(err))
			}
		}()
	}
}

func (t *tokensUseCase) updateAssetsAtHeightIntervalSync(height uint64) error {
	if height%uint64(t.updateAssetsHeightInterval) == 0 {
		if err := t.tokenLoader.FetchAndUpdateTokens(t.LoadTokens); err != nil {
			return err
		}
	}
	return nil
}
```

Note that by having a separate sync version and returning an error, we improve the testability of the method, allowing for the ability to set up table-driven test cases and validate all branches.

Also, please note my TODO about the Prometheus metric

Reply:
Note that this ignores one of my previous comments about initializing `t.tokenLoader` with `t.LoadTokens` rather than propagating it for every call.

Reply:
Totally agree, that simplies a lot of the logic in tests as well. Thanks for suggestion, implemened and pushed relevant changes!